Content.Client/GameObjects/Components/StackVisualizer.cs

```diff
@@ -0,0 +1,145 @@
+            {
+                _isTransparent = transparent.AsBool();
+            }
+        }
+
+        public override void InitializeEntity(IEntity entity)
+        {
+            base.InitializeEntity(entity);
+
+            if (_isTransparent
+                && _spriteLayers.Count > 0
+                && entity.TryGetComponent<ISpriteComponent>(out var spriteComponent))
+            {
+                foreach (var sprite in _spriteLayers)
+                {
+                    var rsiPath = spriteComponent.BaseRSI!.Path!;
+                    spriteComponent.LayerMapReserveBlank(sprite);
+                    spriteComponent.LayerSetSprite(sprite, new SpriteSpecifier.Rsi(rsiPath, sprite));
+                    spriteComponent.LayerSetVisible(sprite, false);
+                }
+            }
+        }
+
+        public override void OnChangeData(AppearanceComponent component)
+        {
+            base.OnChangeData(component);
+
+            if (component.Owner.TryGetComponent<ISpriteComponent>(out var spriteComponent))
+            {
+                if (_isTransparent)
+                {
+                    ProcessTransparentSprites(component, spriteComponent);
+                }
+                else
+                {
+                    ProcessOpaqueSprites(component, spriteComponent);
+                }
+            }
+        }
+
+        private void ProcessOpaqueSprites(AppearanceComponent component, ISpriteComponent spriteComponent)
+        {
+            // Skip processing if no actual/maxCount
+            if (!component.TryGetData<int>(StackVisuals.Actual, out var actual)) return;
+            if (!component.TryGetData<int>(StackVisuals.MaxCount, out var maxCount)) return;
+
+            var activeLayer = ContentHelpers.RoundToNearestIndex(actual, maxCount, _spriteLayers.Count);
+            spriteComponent.LayerSetState(IconLayer, _spriteLayers[activeLayer]);
+        }
+
+        private void ProcessTransparentSprites(AppearanceComponent component, ISpriteComponent spriteComponent)
+        {
+            // If hidden, don't render any sprites
+            if (!component.TryGetData<bool>(StackVisuals.Hide, out var hide)
+                || hide)
+            {
+                foreach (var transparentSprite in _spriteLayers)
+                {
+                    spriteComponent.LayerSetVisible(transparentSprite, false);
+                }
+
+                return;
+            }
+
+            // Skip processing if no actual/maxCount
+            if (!component.TryGetData<int>(StackVisuals.Actual, out var actual)) return;
+            if (!component.TryGetData<int>(StackVisuals.MaxCount, out var maxCount)) return;
+
+            var activeTill = ContentHelpers.RoundToNearestIndex(actual, maxCount, _spriteLayers.Count + 1);
+            for (var i = 0; i < _spriteLayers.Count; i++)
+            {
+                spriteComponent.LayerSetVisible(_spriteLayers[i], i < activeTill);
+            }
```

Comment:
Couldn't you just cut the update loop short if activeTill >= i?

Reply:
I'll look into it, but I believe skipping the visualization  for `activeTill >= i` would lead to problems when you go from `(size-1)` inventory to `size`.